Added cosign plugin in the dynamic buildkite template

hnadiminti-equinix opened the pull request:
Desc:
Cosign plugin, Docker metadata, Docker build plugin, ssm-plugins has been added via changes to "conf.yaml" and changes to template file. On top of it, now with a single command a user should be able to generate buildkite template for multiple plugins. The defaults should come from "conf.yaml" and the overrides for various plugins can be passed as yaml path as below. 


O/P:-
```
Hari.Nadiminti@EUMBQTHV267VXT dynamic-buildkite-template % go run main.go
steps:
  - label: ":docker: get cosign key"
    key: "getkey"
    command: |
      #!/bin/bash
      echo "\$COSIGN_KEY_SECRET" > cosign.key
    plugins:
      - equinixmetal-buildkite/docker-build#v1.1.0:
          build-args:
            - NAME=REPO_NAME
          push: true
      - ssh://git@github.com/equinixmetal/ssm-buildkite-plugin#v1.0.4:
          parameters:
            COSIGN_KEY_SECRET : test-secret
            COSIGN_PASSWORD : passwd
      - equinixmetal-buildkite/cosign#main:
          image: ghcr.io/my-project/my-image:latest
          keyless : false
          keyed-config:
            key: cosign.key
          cosign-version: main
      - equinixmetal-buildkite/trivy#v1.18.3:
          exit-code: 0
          timeout : "5m0s"
          severity: "HIGH,CRITICAL"
          ignore-unfixed: true
          security-checks: "vuln,config"
          skip-files: ""
          skip-dirs: ""
          image-ref: ""
      - equinixmetal-buildkite/docker-metadata#v1.0.0:
          images:
          - "my-org/my-image"
          - "image2"
          extra_tags:
          - "latest"
          - "tag2"                                                                                                               
Hari.Nadiminti@EUMBQTHV267VXT dynamic-buildkite-template % go run main.go --overrides plugins.trivy.skip-files="x.txt,y.txt" --overrides plugins.cosign.keyless=false
steps:
  - label: ":docker: get cosign key"
    key: "getkey"
    command: |
      #!/bin/bash
      echo "\$COSIGN_KEY_SECRET" > cosign.key
    plugins:
      - equinixmetal-buildkite/docker-build#v1.1.0:
          build-args:
            - NAME=REPO_NAME
          push: true
      - ssh://git@github.com/equinixmetal/ssm-buildkite-plugin#v1.0.4:
          parameters:
            COSIGN_KEY_SECRET : test-secret
            COSIGN_PASSWORD : passwd
      - equinixmetal-buildkite/cosign#main:
          image: ghcr.io/my-project/my-image:latest
          keyless : false
          keyed-config:
            key: cosign.key
          cosign-version: main
      - equinixmetal-buildkite/trivy#v1.18.3:
          exit-code: 0
          timeout : "5m0s"
          severity: "HIGH,CRITICAL"
          ignore-unfixed: true
          security-checks: "vuln,config"
          skip-files: "x.txt,y.txt"
          skip-dirs: ""
          image-ref: ""
      - equinixmetal-buildkite/docker-metadata#v1.0.0:
          images:
          - "my-org/my-image"
          - "image2"
          extra_tags:
          - "latest"
          - "tag2"

Hari.Nadiminti@EUMBQTHV267VXT dynamic-buildkite-template % go test -v ./...
?       dynamic-buildkite-template      [no test files]
?       dynamic-buildkite-template/cmd  [no test files]
?       dynamic-buildkite-template/config       [no test files]
=== RUN   TestGenerateTrivyStep
=== RUN   TestGenerateTrivyStep/success
=== RUN   TestGenerateTrivyStep/wrong_template_path
--- PASS: TestGenerateTrivyStep (0.00s)
    --- PASS: TestGenerateTrivyStep/success (0.00s)
    --- PASS: TestGenerateTrivyStep/wrong_template_path (0.00s)
=== RUN   TestGenerateDockerMetadataStep
=== RUN   TestGenerateDockerMetadataStep/success
--- PASS: TestGenerateDockerMetadataStep (0.00s)
    --- PASS: TestGenerateDockerMetadataStep/success (0.00s)
=== RUN   TestSSMStep
=== RUN   TestSSMStep/success
--- PASS: TestSSMStep (0.00s)
    --- PASS: TestSSMStep/success (0.00s)
=== RUN   TestCosignStep
=== RUN   TestCosignStep/success
--- PASS: TestCosignStep (0.00s)
    --- PASS: TestCosignStep/success (0.00s)
PASS
ok      dynamic-buildkite-template/generator    0.604s
=== RUN   TestGetLatestTag
=== RUN   TestGetLatestTag/PublicRepo
=== RUN   TestGetLatestTag/PublicRepo/ValidPAT
=== RUN   TestGetLatestTag/PublicRepo/InvalidPAT
=== RUN   TestGetLatestTag/PublicRepo/NoPAT
=== RUN   TestGetLatestTag/PrivateRepo
=== RUN   TestGetLatestTag/PrivateRepo/ValidPAT
=== RUN   TestGetLatestTag/PrivateRepo/InvalidPAT
=== RUN   TestGetLatestTag/PrivateRepo/NoPAT
--- PASS: TestGetLatestTag (0.95s)
    --- PASS: TestGetLatestTag/PublicRepo (0.71s)
        --- PASS: TestGetLatestTag/PublicRepo/ValidPAT (0.35s)
        --- PASS: TestGetLatestTag/PublicRepo/InvalidPAT (0.15s)
        --- PASS: TestGetLatestTag/PublicRepo/NoPAT (0.20s)
    --- PASS: TestGetLatestTag/PrivateRepo (0.24s)
        --- PASS: TestGetLatestTag/PrivateRepo/ValidPAT (0.03s)
        --- PASS: TestGetLatestTag/PrivateRepo/InvalidPAT (0.17s)
        --- PASS: TestGetLatestTag/PrivateRepo/NoPAT (0.04s)
PASS
ok      dynamic-buildkite-template/util (cached)
``` 
[JIRA ticket](https://packet.atlassian.net/browse/SCT-381)

Buildkite Job log: https://buildkite.com/equinix/buildkite-actions-sandbox/builds/184

prezha commented:
@hnadiminti-equinix, with this pr, we're removing all the flags with the exception of `--config` and `--overrides`?
that's actually done in commit 72700a314854c487e6e596cf5cbe7977c47e8f01: https://github.com/equinixmetal-buildkite/dynamic-buildkite-template/pull/20/commits/72700a314854c487e6e596cf5cbe7977c47e8f01#diff-688f857fa41b86c2e3e022a69cd3a3c1f8b8b47b0d8c57abba0725613d293f87L27-L36
and so we would relay and also require a presence of the config file (ie, will not work without it)
not saying this is wrong, it's actually more simple, but curious if that was the original intention with this tool?

prezha commented on cmd/generate.go:
```diff
@@ -15,6 +18,15 @@ var (
+	generateCmd.Flags().StringToString("overrides", nil, `pass the overrides in the maps syntax as --overrides plugins.trivy.skip-files="x.txt,y.txt" --overrides plugins.cosign.keyless=false`)
+
+	generateCmd.PersistentFlags().StringVar(&ConfigFilePath, "config", "", fmt.Sprintf("config file path (default %q)", defaultConfigFilePath))
+	// log.Println("cosign enabled:", *cosignEnabled)
```
probably a left-over from manual testing

hnadiminti-equinix replied:
Done

prezha commented on cmd/generate.go:
```diff
@@ -23,14 +35,17 @@ Usage of dynamic-buildkite-template
+		// load cosign plugin config
+		LoadCosignConfigs()
+		// generate the build template
+		generator.GenerateBuildSteps(g, os.Stdout, util.TEMPLATE_FILE_PATH)
```
the return value (error) should be checked

hnadiminti-equinix replied:
Done

prezha commented on cmd/generate.go:
```diff
@@ -42,10 +57,25 @@ func initConfig() {
+		vNew.Set(k, v)
+	}
+
+	viper.MergeConfigMap(vNew.AllSettings()) // merge to global viper object
```
the return value (error) should be checked

hnadiminti-equinix replied:
Done

prezha commented on cmd/util.go:
```diff
@@ -74,14 +74,13 @@ func mustGetIntFlag(cmd *cobra.Command, name string) int {
 	return flagVal
 }
```
if most flags are to be removed, then these funcs are not used anymore and should be removed as well

hnadiminti-equinix replied:
Done

prezha commented on util/constants.go:
```diff
@@ -0,0 +1,5 @@
+package util
+
+const (
+	TEMPLATE_FILE_PATH string = "templates/plugins-step.tmpl"
```
nit: a good practice is to use the CamelCase instead of CAPS for var names

hnadiminti-equinix replied:
Done

hnadiminti-equinix commented:
> @hnadiminti-equinix, with this pr, we're removing all the flags with the exception of `--config` and `--overrides`? that's actually done in commit [72700a3](https://github.com/equinixmetal-buildkite/dynamic-buildkite-template/commit/72700a314854c487e6e596cf5cbe7977c47e8f01): [72700a3#diff-688f857fa41b86c2e3e022a69cd3a3c1f8b8b47b0d8c57abba0725613d293f87L27-L36](https://github.com/equinixmetal-buildkite/dynamic-buildkite-template/commit/72700a314854c487e6e596cf5cbe7977c47e8f01#diff-688f857fa41b86c2e3e022a69cd3a3c1f8b8b47b0d8c57abba0725613d293f87L27-L36) and so we would relay and also require a presence of the config file (ie, will not work without it) not saying this is wrong, it's actually more simple, but curious if that was the original intention with this tool?

Hi @prezha , Morning! "Cobra does not allow two sub commands to run at the same time for different plugins. Using the ```conf.yaml``` to have defaults makes it simple to use multiple plugins at the same time, still allowing the user to specify overrides for each plugin in the same command"

prezha commented:
looks like there's are some issue with eg, `keyless` - ie, when keyless is `true` (as is - by default, but also when explicitly set), it does not appear in the config, but `keyless-config` does:
```
$ go run .
steps:
  - command: ls
    plugins:
      - equinixmetal-buildkite/cosign#v0.1.0:
          image: ghcr.io/my-project/my-image:latest
          keyless-config:
            fulcio-url: https://fulcio.sigstore.dev
            rekor-url: https://rekor.sigstore.dev
          cosign-version: v0.1.0
      - equinixmetal-buildkite/trivy#v1.18.3:
          exit-code: 0
          timeout : "5m0s"
          severity: "HIGH,CRITICAL"
          ignore-unfixed: true
          security-checks: "vuln,config"
          skip-files: ""
          skip-dirs: ""
          image-ref: ""%
```
```
go run main.go --overrides plugins.trivy.skip-files="x.txt,y.txt" --overrides plugins.cosign.keyless=true
steps:
  - command: ls
    plugins:
      - equinixmetal-buildkite/cosign#v0.1.0:
          image: ghcr.io/my-project/my-image:latest
          keyless-config:
            fulcio-url: https://fulcio.sigstore.dev
            rekor-url: https://rekor.sigstore.dev
          cosign-version: v0.1.0
      - equinixmetal-buildkite/trivy#v1.18.3:
          exit-code: 0
          timeout : "5m0s"
          severity: "HIGH,CRITICAL"
          ignore-unfixed: true
          security-checks: "vuln,config"
          skip-files: "x.txt,y.txt"
          skip-dirs: ""
          image-ref: ""%
```
i haven't checked other combinations, would be good to test them

prezha commented:
hi @hnadiminti-equinix looks like some tests are failing, can you please have a look
```
$ go test -v ./...
?   	dynamic-buildkite-template	[no test files]
?   	dynamic-buildkite-template/cmd	[no test files]
?   	dynamic-buildkite-template/config	[no test files]
=== RUN   TestGenerateTrivyStep
=== RUN   TestGenerateTrivyStep/success
=== RUN   TestGenerateTrivyStep/wrong_template_path
--- PASS: TestGenerateTrivyStep (0.00s)
    --- PASS: TestGenerateTrivyStep/success (0.00s)
    --- PASS: TestGenerateTrivyStep/wrong_template_path (0.00s)
=== RUN   TestGenerateDockerMetadataStep
=== RUN   TestGenerateDockerMetadataStep/success
    common-pipeline_test.go:121: Not equal:
        expected: steps:
          - command: ls
            plugins:
              - equinixmetal-buildkite/docker-metadata#:
                  images: "[test_image]"
                  title: "test_title"
                  vendor: "test_vendor"
        actual  : steps:
          - command: ls
            plugins:
              - equinixmetal-buildkite/docker-metadata#:
                  images:
                  - "test_image"
                  title: "test_title"
                  vendor: "test_vendor"
=== RUN   TestGenerateDockerMetadataStep/wrong_template_path
--- FAIL: TestGenerateDockerMetadataStep (0.00s)
    --- FAIL: TestGenerateDockerMetadataStep/success (0.00s)
    --- PASS: TestGenerateDockerMetadataStep/wrong_template_path (0.00s)
FAIL
FAIL	dynamic-buildkite-template/generator	0.121s
=== RUN   TestGetLatestTag
=== RUN   TestGetLatestTag/PublicRepo
=== RUN   TestGetLatestTag/PublicRepo/ValidPAT
=== RUN   TestGetLatestTag/PublicRepo/InvalidPAT
=== RUN   TestGetLatestTag/PublicRepo/NoPAT
=== RUN   TestGetLatestTag/PrivateRepo
=== RUN   TestGetLatestTag/PrivateRepo/ValidPAT
=== RUN   TestGetLatestTag/PrivateRepo/InvalidPAT
=== RUN   TestGetLatestTag/PrivateRepo/NoPAT
--- PASS: TestGetLatestTag (1.27s)
    --- PASS: TestGetLatestTag/PublicRepo (0.65s)
        --- PASS: TestGetLatestTag/PublicRepo/ValidPAT (0.30s)
        --- PASS: TestGetLatestTag/PublicRepo/InvalidPAT (0.15s)
        --- PASS: TestGetLatestTag/PublicRepo/NoPAT (0.21s)
    --- PASS: TestGetLatestTag/PrivateRepo (0.63s)
        --- PASS: TestGetLatestTag/PrivateRepo/ValidPAT (0.31s)
        --- PASS: TestGetLatestTag/PrivateRepo/InvalidPAT (0.13s)
        --- PASS: TestGetLatestTag/PrivateRepo/NoPAT (0.18s)
PASS
ok  	dynamic-buildkite-template/util	1.410s
FAIL
```

prezha commented on conf.yaml:
```diff
@@ -2,11 +2,30 @@ plugins:
+    helm-overrides-files: ""
+  cosign:
+    image: "ghcr.io/my-project/my-image:latest"
+    keyless: true
```
we would want to be careful about this not being the default - ie, for private/internal repos we'd use private/internal sigstore instance

hnadiminti-equinix replied:
Addressed, I updated and added keyless=false as a default.

hnadiminti-equinix commented:
> hi @hnadiminti-equinix looks like some tests are failing, can you please have a look
> 
> ```
> $ go test -v ./...
> ?   	dynamic-buildkite-template	[no test files]
> ?   	dynamic-buildkite-template/cmd	[no test files]
> ?   	dynamic-buildkite-template/config	[no test files]
> === RUN   TestGenerateTrivyStep
> === RUN   TestGenerateTrivyStep/success
> === RUN   TestGenerateTrivyStep/wrong_template_path
> --- PASS: TestGenerateTrivyStep (0.00s)
>     --- PASS: TestGenerateTrivyStep/success (0.00s)
>     --- PASS: TestGenerateTrivyStep/wrong_template_path (0.00s)
> === RUN   TestGenerateDockerMetadataStep
> === RUN   TestGenerateDockerMetadataStep/success
>     common-pipeline_test.go:121: Not equal:
>         expected: steps:
>           - command: ls
>             plugins:
>               - equinixmetal-buildkite/docker-metadata#:
>                   images: "[test_image]"
>                   title: "test_title"
>                   vendor: "test_vendor"
>         actual  : steps:
>           - command: ls
>             plugins:
>               - equinixmetal-buildkite/docker-metadata#:
>                   images:
>                   - "test_image"
>                   title: "test_title"
>                   vendor: "test_vendor"
> === RUN   TestGenerateDockerMetadataStep/wrong_template_path
> --- FAIL: TestGenerateDockerMetadataStep (0.00s)
>     --- FAIL: TestGenerateDockerMetadataStep/success (0.00s)
>     --- PASS: TestGenerateDockerMetadataStep/wrong_template_path (0.00s)
> FAIL
> FAIL	dynamic-buildkite-template/generator	0.121s
> === RUN   TestGetLatestTag
> === RUN   TestGetLatestTag/PublicRepo
> === RUN   TestGetLatestTag/PublicRepo/ValidPAT
> === RUN   TestGetLatestTag/PublicRepo/InvalidPAT
> === RUN   TestGetLatestTag/PublicRepo/NoPAT
> === RUN   TestGetLatestTag/PrivateRepo
> === RUN   TestGetLatestTag/PrivateRepo/ValidPAT
> === RUN   TestGetLatestTag/PrivateRepo/InvalidPAT
> === RUN   TestGetLatestTag/PrivateRepo/NoPAT
> --- PASS: TestGetLatestTag (1.27s)
>     --- PASS: TestGetLatestTag/PublicRepo (0.65s)
>         --- PASS: TestGetLatestTag/PublicRepo/ValidPAT (0.30s)
>         --- PASS: TestGetLatestTag/PublicRepo/InvalidPAT (0.15s)
>         --- PASS: TestGetLatestTag/PublicRepo/NoPAT (0.21s)
>     --- PASS: TestGetLatestTag/PrivateRepo (0.63s)
>         --- PASS: TestGetLatestTag/PrivateRepo/ValidPAT (0.31s)
>         --- PASS: TestGetLatestTag/PrivateRepo/InvalidPAT (0.13s)
>         --- PASS: TestGetLatestTag/PrivateRepo/NoPAT (0.18s)
> PASS
> ok  	dynamic-buildkite-template/util	1.410s
> FAIL
> ```

That was my bad...it's now fixed and all test cases are being passed.
Can you please check...Thanks @prezha

oliver-equinix commented on templates/plugins-step.tmpl:
```diff
@@ -0,0 +1,99 @@
+    plugins:
+{{- end }}
+{{- if .SSMPluginEnabled }}
+      - ssh://git@git@github.com/equinixmetal/ssm-buildkite-plugin#{{ .CosignConfig.CosignVersion }}:
```
1 too many `git@`s in the url

hnadiminti-equinix replied:
Thanks for the review @oliver-equinix, the change has been addressed..

prezha commented:
if run as-is, i think it still might have several issues - eg:

- where will the `COSIGN_KEY_SECRET` be stored, ie, the `COSIGN_KEY_PATH` is not defined?
- `COSIGN_KEY_SECRET : test-secret`, `COSIGN_PASSWORD : passwd`, `GITHUB_TOKEN : token`, `key: sample-key` under the `keyed-config`, `image: ghcr.io/my-project/my-image:latest`, etc. would not work - ie, should be parameterised, not generic example constant values
- when defaults are used, we have:
```
$ go run .
steps:
  - label: ":docker: get cosign key"
    key: "getkey"
    command: |
      #!/bin/bash
      echo "\$COSIGN_KEY_SECRET" > ${COSIGN_KEY_PATH}
    plugins:
      - ssh://git@github.com/equinixmetal/ssm-buildkite-plugin#v0.1.0:
          parameters:
            COSIGN_KEY_SECRET : test-secret
            COSIGN_PASSWORD : passwd
            GITHUB_TOKEN : token
      - equinixmetal-buildkite/cosign#v0.1.0:
          image: ghcr.io/my-project/my-image:latest
          keyless : false
          keyed-config:
            key: sample-key
          cosign-version: v0.1.0
      - equinixmetal-buildkite/trivy#v1.18.3:
          exit-code: 0
          timeout : "5m0s"
          severity: "HIGH,CRITICAL"
          ignore-unfixed: true
          security-checks: "vuln,config"
          skip-files: ""
          skip-dirs: ""
          image-ref: ""
      - equinixmetal-buildkite/docker-metadata#v1.0.0:
          images:
          - "my-org/my-image"
          - "image2"
          extra_tags:
          - "latest"
          - "tag2"
          title: "sample"
          vendor: "abc"
          debug: "true"
```
then, if user wants to use keyless for cosign, the `keyless: true` param is missing from the output yaml conf (could be considered the default), but also the `keyless-config` is still present and `COSIGN_KEY_SECRET` and `COSIGN_PASSWORD` are still "pulled" from ssm, which should not be needed (ie, only needed for the keyed and not for the keyless signing)? eg:
```
go run main.go --overrides plugins.cosign.keyless=true
steps:
  - label: ":docker: get cosign key"
    key: "getkey"
    command: |
      #!/bin/bash
      echo "\$COSIGN_KEY_SECRET" > ${COSIGN_KEY_PATH}
    plugins:
      - ssh://git@github.com/equinixmetal/ssm-buildkite-plugin#v0.1.0:
          parameters:
            COSIGN_KEY_SECRET : test-secret
            COSIGN_PASSWORD : passwd
            GITHUB_TOKEN : token
      - equinixmetal-buildkite/cosign#v0.1.0:
          image: ghcr.io/my-project/my-image:latest
          keyless-config:
            fulcio-url: https://fulcio.sigstore.dev
            rekor-url: https://rekor.sigstore.dev
          cosign-version: v0.1.0
      - equinixmetal-buildkite/trivy#v1.18.3:
          exit-code: 0
          timeout : "5m0s"
          severity: "HIGH,CRITICAL"
          ignore-unfixed: true
          security-checks: "vuln,config"
          skip-files: ""
          skip-dirs: ""
          image-ref: ""
      - equinixmetal-buildkite/docker-metadata#v1.0.0:
          images:
          - "my-org/my-image"
          - "image2"
          extra_tags:
          - "latest"
          - "tag2"
          title: "sample"
          vendor: "abc"
          debug: "true"
```

hnadiminti-equinix commented:
> if run as-is, i think it still might have several issues - eg:
> 
> * where will the `COSIGN_KEY_SECRET` be stored, ie, the `COSIGN_KEY_PATH` is not defined?
> * `COSIGN_KEY_SECRET : test-secret`, `COSIGN_PASSWORD : passwd`, `GITHUB_TOKEN : token`, `key: sample-key` under the `keyed-config`, `image: ghcr.io/my-project/my-image:latest`, etc. would not work - ie, should be parameterised, not generic example constant values?
> * what the `GITHUB_TOKEN` is used for here, ie, why we have it in the generated template?
> * when defaults are used, we have:
> 
> ```
> $ go run .
> steps:
>   - label: ":docker: get cosign key"
>     key: "getkey"
>     command: |
>       #!/bin/bash
>       echo "\$COSIGN_KEY_SECRET" > ${COSIGN_KEY_PATH}
>     plugins:
>       - ssh://git@github.com/equinixmetal/ssm-buildkite-plugin#v0.1.0:
>           parameters:
>             COSIGN_KEY_SECRET : test-secret
>             COSIGN_PASSWORD : passwd
>             GITHUB_TOKEN : token
>       - equinixmetal-buildkite/cosign#v0.1.0:
>           image: ghcr.io/my-project/my-image:latest
>           keyless : false
>           keyed-config:
>             key: sample-key
>           cosign-version: v0.1.0
>       - equinixmetal-buildkite/trivy#v1.18.3:
>           exit-code: 0
>           timeout : "5m0s"
>           severity: "HIGH,CRITICAL"
>           ignore-unfixed: true
>           security-checks: "vuln,config"
>           skip-files: ""
>           skip-dirs: ""
>           image-ref: ""
>       - equinixmetal-buildkite/docker-metadata#v1.0.0:
>           images:
>           - "my-org/my-image"
>           - "image2"
>           extra_tags:
>           - "latest"
>           - "tag2"
>           title: "sample"
>           vendor: "abc"
>           debug: "true"
> ```
> 
> then, if user wants to use keyless for cosign, the `keyless: true` param is missing from the output yaml conf (could be considered the default), but also the `keyless-config` is still present and `COSIGN_KEY_SECRET` and `COSIGN_PASSWORD` are still "pulled" from ssm, which should not be needed (ie, only needed for the keyed and not for the keyless signing)? eg:
> 
> ```
> go run main.go --overrides plugins.cosign.keyless=true
> steps:
>   - label: ":docker: get cosign key"
>     key: "getkey"
>     command: |
>       #!/bin/bash
>       echo "\$COSIGN_KEY_SECRET" > ${COSIGN_KEY_PATH}
>     plugins:
>       - ssh://git@github.com/equinixmetal/ssm-buildkite-plugin#v0.1.0:
>           parameters:
>             COSIGN_KEY_SECRET : test-secret
>             COSIGN_PASSWORD : passwd
>             GITHUB_TOKEN : token
>       - equinixmetal-buildkite/cosign#v0.1.0:
>           image: ghcr.io/my-project/my-image:latest
>           keyless-config:
>             fulcio-url: https://fulcio.sigstore.dev
>             rekor-url: https://rekor.sigstore.dev
>           cosign-version: v0.1.0
>       - equinixmetal-buildkite/trivy#v1.18.3:
>           exit-code: 0
>           timeout : "5m0s"
>           severity: "HIGH,CRITICAL"
>           ignore-unfixed: true
>           security-checks: "vuln,config"
>           skip-files: ""
>           skip-dirs: ""
>           image-ref: ""
>       - equinixmetal-buildkite/docker-metadata#v1.0.0:
>           images:
>           - "my-org/my-image"
>           - "image2"
>           extra_tags:
>           - "latest"
>           - "tag2"
>           title: "sample"
>           vendor: "abc"
>           debug: "true"
> ```



> if run as-is, i think it still might have several issues - eg:
> 
> * where will the `COSIGN_KEY_SECRET` be stored, ie, the `COSIGN_KEY_PATH` is not defined?
> * `COSIGN_KEY_SECRET : test-secret`, `COSIGN_PASSWORD : passwd`, `GITHUB_TOKEN : token`, `key: sample-key` under the `keyed-config`, `image: ghcr.io/my-project/my-image:latest`, etc. would not work - ie, should be parameterised, not generic example constant values?
> * what the `GITHUB_TOKEN` is used for here, ie, why we have it in the generated template?
> * when defaults are used, we have:
> 
> ```
> $ go run .
> steps:
>   - label: ":docker: get cosign key"
>     key: "getkey"
>     command: |
>       #!/bin/bash
>       echo "\$COSIGN_KEY_SECRET" > ${COSIGN_KEY_PATH}
>     plugins:
>       - ssh://git@github.com/equinixmetal/ssm-buildkite-plugin#v0.1.0:
>           parameters:
>             COSIGN_KEY_SECRET : test-secret
>             COSIGN_PASSWORD : passwd
>             GITHUB_TOKEN : token
>       - equinixmetal-buildkite/cosign#v0.1.0:
>           image: ghcr.io/my-project/my-image:latest
>           keyless : false
>           keyed-config:
>             key: sample-key
>           cosign-version: v0.1.0
>       - equinixmetal-buildkite/trivy#v1.18.3:
>           exit-code: 0
>           timeout : "5m0s"
>           severity: "HIGH,CRITICAL"
>           ignore-unfixed: true
>           security-checks: "vuln,config"
>           skip-files: ""
>           skip-dirs: ""
>           image-ref: ""
>       - equinixmetal-buildkite/docker-metadata#v1.0.0:
>           images:
>           - "my-org/my-image"
>           - "image2"
>           extra_tags:
>           - "latest"
>           - "tag2"
>           title: "sample"
>           vendor: "abc"
>           debug: "true"
> ```
> 
> then, if user wants to use keyless for cosign, the `keyless: true` param is missing from the output yaml conf (could be considered the default), but also the `keyless-config` is still present and `COSIGN_KEY_SECRET` and `COSIGN_PASSWORD` are still "pulled" from ssm, which should not be needed (ie, only needed for the keyed and not for the keyless signing)? eg:
> 
> ```
> go run main.go --overrides plugins.cosign.keyless=true
> steps:
>   - label: ":docker: get cosign key"
>     key: "getkey"
>     command: |
>       #!/bin/bash
>       echo "\$COSIGN_KEY_SECRET" > ${COSIGN_KEY_PATH}
>     plugins:
>       - ssh://git@github.com/equinixmetal/ssm-buildkite-plugin#v0.1.0:
>           parameters:
>             COSIGN_KEY_SECRET : test-secret
>             COSIGN_PASSWORD : passwd
>             GITHUB_TOKEN : token
>       - equinixmetal-buildkite/cosign#v0.1.0:
>           image: ghcr.io/my-project/my-image:latest
>           keyless-config:
>             fulcio-url: https://fulcio.sigstore.dev
>             rekor-url: https://rekor.sigstore.dev
>           cosign-version: v0.1.0
>       - equinixmetal-buildkite/trivy#v1.18.3:
>           exit-code: 0
>           timeout : "5m0s"
>           severity: "HIGH,CRITICAL"
>           ignore-unfixed: true
>           security-checks: "vuln,config"
>           skip-files: ""
>           skip-dirs: ""
>           image-ref: ""
>       - equinixmetal-buildkite/docker-metadata#v1.0.0:
>           images:
>           - "my-org/my-image"
>           - "image2"
>           extra_tags:
>           - "latest"
>           - "tag2"
>           title: "sample"
>           vendor: "abc"
>           debug: "true"
> ```

I fixed and updated the code; could you perhaps verify and let me know. The parameterisation will be updated after I test the job in the buildkite CI pipeline. Thank you!
